mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,290 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Module contains Node and Assertion classes which together create a local
+ * version of the Data Commons Knowledge Graph.
+ */
+
+import {
+  doesExistsInKG,
+  getRemotePropertyLabels,
+  getRemotePropertyValues,
+  getValueFromValueObj,
+} from './utils.js';
+
+/** Class representation of a single Node in the KG. */
+class Node {
+  /**
+   * The local id used in a parsed mcf file.
+   * @type {string}
+   */
+  localId;
+
+  /**
+   * Whether triples from the remote Data Commons Knowledge Graph have already
+   * been fetched.
+   * @type {boolean}
+   */
+
+  alreadyFetched;
+
+  /**
+   * Whether the node exists in the Data Commons Knowledge Graph.
+   * @type {boolean}
+   */
+  existsInKG;
+
+  /**
+   * Dcid of the node. Set only if a remote id is referred to in a local file or
+   * pulled from the Data Commons Knowledge Graph.
+   * @type {string}
+   */
+
+  dcid;
+  /**
+   * Array of Assertion objects representing the outgoing triples of the Node
+   * object.
+   * @type {Array<Assertion>}
+   */
+  assertions;
+  /**
+   * Array of Assertion objects representing the incoming triples of the Node
+   * object.
+   * @type {Array<Assertion>}
+   */
+  invAssertions;
+
+  /**
+   * Create a Node based on a given id.
+   * @param {string} id The id of the node to create, including the namespace.
+   */
+  constructor(id) {
+    this.localId = id.startsWith('l:') ? id : null;
+    this.dcid = null;
+
+    this.alreadyFetched = false;
+    this.existsInKG = false;
+
+    this.assertions = [];
+    this.invAssertions = [];
+
+    Node.nodeHash[id] = this;
+  }
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
```

Comment:
comment needs update?  could you point out why shouldCreate isn't needed anymore (i.e., are there no call sites that lookup an existing node)?